### PR TITLE
Serialize more dtypes

### PR DESCRIPTION
**Pull request by pitrou**
Structured dtypes (kind 'V') can have additional complication which require more care than is currently done.

Also remove unnecessary `stride_tricks` calls.

### CODE REVIEW

**Comment by pitrou** [comment]
@mrocklin

File: distributed/protocol/numpy.py [hunk 1]
```diff
@@ -33,22 +32,36 @@ def serialize_numpy_ndarray(x):
         frames = [pickle.dumps(x)]
         return header, frames
 
+    # We cannot blindly pickle the dtype as some may fail pickling,
+    # so we have a mixture of strategies.
```
Comment (mrocklin) [inline distributed/protocol/numpy.py]: Does cloudpickle work in all cases?  (might be simpler if so)

Comment (pitrou) [inline distributed/protocol/numpy.py]: cloudpickle doesn't have specific code for dtypes, so I doubt so.

Comment (mrocklin) [inline distributed/protocol/numpy.py]: Should we also raise an issue upstream in numpy?

File: distributed/protocol/numpy.py [hunk 2]
```diff
@@ -33,22 +32,36 @@ def serialize_numpy_ndarray(x):
...
+        # Preserving all the information works best when pickling
+        try:
+            # Only use stdlib pickle as cloudpickle is slow when failing
+            # (microseconds instead of nanoseconds)
+            dt = (1, pickle.pickle.dumps(x.dtype))
+            pickle.loads(dt[1])  # does it unpickle fine?
+        except Exception:
+            # dtype fails pickling => fall back on the descr if reasonable.
+            if x.dtype.type is not np.void or x.dtype.alignment != 1:
+                raise
```
Comment (mrocklin) [inline distributed/protocol/numpy.py]: Does this occur?  Are there known cases where our serialization fails?

File: distributed/protocol/numpy.py [hunk 3]
```diff
@@ -33,22 +32,36 @@ def serialize_numpy_ndarray(x):
...
+            else:
+                dt = (0, x.dtype.descr)
     else:
-        dt = x.dtype.str
+        dt = (0, x.dtype.str)
```
Comment (mrocklin) [inline distributed/protocol/numpy.py]: Is this effectively a boolean?  If so should we use a bool?  Or is the thought here that we might have to have other schemes in the future?

**Comment by mrocklin** [comment]
I'm generally +1 on this.  I have some questions, but they shouldn't stop merging.

File: distributed/protocol/numpy.py (hunk 1 above)
Comment (pitrou) [inline distributed/protocol/numpy.py]: https://github.com/numpy/numpy/issues/9159

File: distributed/protocol/numpy.py (hunk 2 above)
Comment (pitrou) [inline distributed/protocol/numpy.py]: Yes, for example: `np.dtype([('a', 'i1'), ('b', 'm8')], align=True)`

File: distributed/protocol/numpy.py (hunk 3 above)
Comment (pitrou) [inline distributed/protocol/numpy.py]: Yes, it's a boolean. I don't think it matters much but we can switch to a boolean if you prefer.

**Pull request merged**